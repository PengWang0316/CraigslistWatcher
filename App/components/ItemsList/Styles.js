import { StyleSheet } from 'react-native';

export default StyleSheet.create({
  separatorStyle: {
    height: 1,
    width: '86%',
    backgroundColor: '#CED0CE',
    marginLeft: '14%'
  },
  indicatorViewStyle: {
    paddingTop: 20,
    paddingBottom: 80,
    borderTopWidth: 1,
    borderColor: '#CED0CE'
  }
});
