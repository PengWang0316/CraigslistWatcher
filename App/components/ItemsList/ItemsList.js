import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { View, ActivityIndicator, FlatList } from 'react-native';
// import { ListItem, Icon } from 'react-native-elements';

import Styles from './Styles';
import { fetchLists, fetchMoreLists } from '../../actions/ListActions';
import { increasePageNumber } from '../../actions/CraigslistStatesActions';
import CraigslistListItem from '../CraigslistListItem';

/** The component are used to show the item list that fetched from CraigsList. */
export class ItemsList extends Component {
  static propTypes = {
    lists: PropTypes.array.isRequired,
    craigslistStates: PropTypes.object.isRequired,
    fetchLists: PropTypes.func.isRequired,
    fetchMoreLists: PropTypes.func.isRequired,
    increasePageNumber: PropTypes.func.isRequired,
    navigation: PropTypes.object.isRequired
  };

  /**
   * If a new Lists data is loaded, set refreshing state to false.
   * @param {object} nextProps contains new prop values.
   * @param {object} prevState contains old state values.
   * @return {object} Return a new state.
   */
  static getDerivedStateFromProps(nextProps, prevState) {
    if (prevState.lists !== nextProps.lists) return { refreshing: false };
    return null;
  }

  state = { refreshing: false };

  /**
   * Fetching the inintal list data from CraigsList.
   * @return {null} No return.
   */
  componentDidMount() {
    this.props.fetchLists({
      city: this.props.craigslistStates.city,
      keyword: this.props.craigslistStates.keyword,
      category: this.props.craigslistStates.category
    });
  }

  /**
   * Loading new data from the data source.
   * @return {null} No return.
   */
  handleRefresh = () => this.setState({ refreshing: true }, () => this.props.fetchLists({
    city: this.props.craigslistStates.city,
    keyword: this.props.craigslistStates.keyword,
    category: this.props.craigslistStates.category
  }));

  /**
   * Loading new data from the data source when a user reach the end of the list.
   * @return {null} No return.
   */
  handleEndReached = () => {
    this.props.fetchMoreLists({
      page: this.props.craigslistStates.pageNumber + 1,
      city: this.props.craigslistStates.city,
      keyword: this.props.craigslistStates.keyword,
      category: this.props.craigslistStates.category
    });
    this.props.increasePageNumber();
  };

  /**
   * Rendering eatch item for the FlatList.
   * @return {jsx} Return a CraigslistListItem component.
   */
  renderItem = ({ item }) => <CraigslistListItem item={item} navigation={this.props.navigation} />;

  /**
   * Rendering the footer for the FlatList.
   * @return {jsx} Return null when lists is empty. Return an activity indicator when the lists is not empty.
   */
  renderFooter = () => {
    if (this.props.lists.length === 0) return null;
    return (
      <View style={Styles.indicatorViewStyle}>
        <ActivityIndicator animating size="large" />
      </View>
    );
  };

  /**
   * Rendering the separator for the FlatList.
   * @return {jsx} Return a view with separator style.
   */
  renderSeparator = () => <View style={Styles.separatorStyle} />;

  /**
   * Render the jsx for the component.
   * @param {object} props contains the prop values for the component.
   * @return {jsx} Return jsx.
   */
  render() {
    const { lists } = this.props;
    return (
      <View>
        {lists.length === 0 && <ActivityIndicator size="large" color="gray" />}
        {lists.length !== 0 &&
        <FlatList
          data={lists}
          keyExtractor={item => item.dataId}
          renderItem={this.renderItem}
          ItemSeparatorComponent={this.renderSeparator}
          ListFooterComponent={this.renderFooter}
          refreshing={this.state.refreshing}
          onRefresh={this.handleRefresh}
          onEndReached={this.handleEndReached}
          onEndThreshold={10}
        />
        }
      </View>
    );
  }
}
/* istanbul ignore next */
const mapStateToProps = state => ({
  lists: state.lists,
  craigslistStates: state.craigslistStates
});
/* istanbul ignore next */
const mapDispatchToProps = dispatch => ({
  fetchLists: parameter => dispatch(fetchLists(parameter)),
  fetchMoreLists: parameter => dispatch(fetchMoreLists(parameter)),
  increasePageNumber: () => dispatch(increasePageNumber())
});
export default connect(mapStateToProps, mapDispatchToProps)(ItemsList);
