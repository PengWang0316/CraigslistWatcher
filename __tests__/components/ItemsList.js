import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';

import { ItemsList } from '../../App/components/ItemsList/ItemsList';

jest.mock('../../App/components/CraigslistListItem', () => 'CraigslistListItem');

describe('Test ItemsList', () => {
  const defaultPorps = {
    lists: [],
    craigslistStates: {
      city: 'city', keyword: 'keyword', category: 'category', pageNumber: 2
    },
    fetchLists: jest.fn(),
    fetchMoreLists: jest.fn(),
    increasePageNumber: jest.fn()
  };

  const getComponent = (props = defaultPorps) => shallow(<ItemsList {...props} />);

  test('initial states and componentDidMount', () => {
    const component = getComponent();
    expect(component.state('refreshing')).toBe(false);
    expect(defaultPorps.fetchLists).toHaveBeenCalledWith({
      city: 'city', keyword: 'keyword', category: 'category'
    });
    expect(defaultPorps.fetchLists).toHaveBeenCalledTimes(1);
  });

  test('getDerivedStateFromProps', () => {
    const component = getComponent();
    const newLists = [];
    component.setState({ refreshing: true });
    component.setProps({ lists: newLists });
    expect(component.state('refreshing')).toBe(false);
    component.setState({ refreshing: true, lists: newLists });
    component.setProps({ lists: newLists });
    expect(component.state('refreshing')).toBe(true);
  });

  test('handleRefresh', () => {
    const component = getComponent({ ...defaultPorps, craigslistStates: { city: 'new city', keyword: 'new keyword', category: 'new category' } });
    component.instance().handleRefresh();
    expect(component.state('refreshing')).toBe(true);
    expect(defaultPorps.fetchLists).toHaveBeenCalledWith({
      city: 'new city', keyword: 'new keyword', category: 'new category'
    });
  });

  test('handleEndReached', () => {
    const component = getComponent();
    component.instance().handleEndReached();
    expect(defaultPorps.fetchMoreLists).toHaveBeenCalledWith({
      city: 'city', keyword: 'keyword', category: 'category', page: 3
    });
    expect(defaultPorps.increasePageNumber).toHaveBeenCalledTimes(1);
  });

  test('renderItem snapshot', () => {
    const component = getComponent();
    expect(renderer.create(component.instance().renderItem({ item: 'item' })).toJSON()).toMatchSnapshot();
  });

  test('renderSeparator snapshot', () => {
    const component = getComponent();
    expect(renderer.create(component.instance().renderSeparator()).toJSON()).toMatchSnapshot();
  });

  test('renderFooter snapshot', () => {
    const component = getComponent();
    expect(component.instance().renderFooter()).toBeNull();
    component.setProps({ lists: [1] });
    expect(renderer.create(component.instance().renderFooter()).toJSON()).toMatchSnapshot();
  });

  test('snapshot test empty lists', () => expect(renderer.create(<ItemsList {...defaultPorps} />).toJSON()).toMatchSnapshot());

  test('snapshot test not empty lists', () => expect(renderer.create(<ItemsList {...{ ...defaultPorps, lists: [1] }} />).toJSON()).toMatchSnapshot());
});
