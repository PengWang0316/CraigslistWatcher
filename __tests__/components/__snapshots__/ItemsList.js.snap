// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test ItemsList renderFooter snapshot 1`] = `
<View
  style={
    Object {
      "borderColor": "#CED0CE",
      "borderTopWidth": 1,
      "paddingBottom": 80,
      "paddingTop": 20,
    }
  }
>
  <ActivityIndicator
    animating={true}
    color="#999999"
    hidesWhenStopped={true}
    size="large"
  />
</View>
`;

exports[`Test ItemsList renderItem snapshot 1`] = `
<CraigslistListItem
  item="item"
/>
`;

exports[`Test ItemsList renderSeparator snapshot 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#CED0CE",
      "height": 1,
      "marginLeft": "14%",
      "width": "86%",
    }
  }
/>
`;

exports[`Test ItemsList snapshot test empty lists 1`] = `
<View>
  <ActivityIndicator
    animating={true}
    color="gray"
    hidesWhenStopped={true}
    size="large"
  />
</View>
`;

exports[`Test ItemsList snapshot test not empty lists 1`] = `<View />`;
